src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorPullDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Diagnostics;
+
+internal class RazorPullDiagnosticsEndpoint : IRazorPullDiagnosticsEndpoint
+{
+    public bool MutatesSolutionState => false;
+
+    public RegistrationExtensionResult GetRegistration(VSInternalClientCapabilities clientCapabilities)
+    {
+        return new RegistrationExtensionResult("_vs_supportsDiagnosticRequests", options: clientCapabilities.SupportsDiagnosticRequests);
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(VSInternalDocumentDiagnosticsParams request)
+    {
+        if (request.TextDocument is null)
+        {
+            throw new NotImplementedException("Why would document be null?");
+        }
+
+        return request.TextDocument;
+    }
+
+    public async Task<IEnumerable<VSInternalDiagnosticReport>?> HandleRequestAsync(VSInternalDocumentDiagnosticsParams request, RazorRequestContext context, CancellationToken cancellationToken)
+    {
+        var result = await GetPullDiagnosticsAsync(request, context, cancellationToken);
+
+        return result;
+    }
+
+    private async Task<IEnumerable<VSInternalDiagnosticReport>?> GetPullDiagnosticsAsync(VSInternalDocumentDiagnosticsParams request, RazorRequestContext requestContext, CancellationToken cancellationToken)
```

Comment:
Rename to `GetHtmlAndCSharpPullDiagnostics` perhaps?

Do Razor diagnostics come from the `RazorTranslateDiagnosticsEndpoint`? Can that all be moved to this endpoint too, so we only have one?